### PR TITLE
Get API request limit and offset from parameters.py

### PR DESCRIPTION
**Pull request by svetli-n**
---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by ephraimbuddy** [comment]
I think it will be better to allow connexion pass it for us by adding it as keyword arguments to the views. What do you think? Take a look at this:
https://github.com/apache/airflow/blob/f6bd817a3aac0a16430fc2e3d59c1f17a69a15ac/airflow/api_connexion/endpoints/event_log_endpoint.py#L42

**Comment by mik-laj** [comment]
Here is doc about it: https://connexion.readthedocs.io/en/latest/request.html#automatic-parameter-handling

**Comment by svetli-n** [comment]
Thanks @mik-laj

@ephraimbuddy you are right, its better. Shall we also add `maximum: 100` to https://github.com/apache/airflow/blob/master/airflow/api_connexion/openapi/v1.yaml#L2135 ?

**Comment by ephraimbuddy** [comment]
No. The default of 100 means that if limit is not applied then a limit of 100 will be applied. Users should be able to fetch more than 100 items if they want

**Comment by svetli-n** [comment]
I believe they can fetch as many as the want by using pagination. Isn't it ddos risk without upper bound?

**Comment by ephraimbuddy** [comment]
@mik-laj

**Comment by OmairK** [comment]
> I believe they can fetch as many as the want by using pagination. Isn't it ddos risk without upper bound?

AFAIK we would be using something like [flask-limiter](https://flask-limiter.readthedocs.io/en/stable/) for rate-limiting our REST-API, which could be configured either globally or for individual routes.
Issue that deals with API security #8113

**Comment by mik-laj** [comment]
@OmairK Rate limit is another problem that can be solved on a different layer, e.g. Load Balancer, so we don't have to deal with it for now.

@ephraimbuddy @svetli-n We should have a limit that should be configured in `airflow.cfg`, but for now we can set it as a fixed value - 100. This is to protect against a DOS attack. Fetching one billion records into memory can cause all server memory to be consumed. This problem will not occur if the user fetches only a subset of items in one request. What do you think about it?

**Comment by svetli-n** [comment]
Yes, those were my thoughts as well (and DOS happens also often by mistake, e.g. typo or bug). I could set max 100 in v1.yaml to avoid all that if no one objects.

**Comment by svetli-n** [comment]
If you are interested in supporting fetching large number of results, may be we can get inspiration from https://www.elastic.co/guide/en/elasticsearch/reference/6.8/search-request-scroll.html, I could research it.

**Comment by svetli-n** [comment]
There are also other pagination options: https://nordicapis.com/everything-you-need-to-know-about-api-pagination

Is it something worth considering?

**Comment by mik-laj** [comment]
> Is it something worth considering?

I think that in the first version of the API we can choose the simplest solution - offset and limit. I analyzed different solutions, but each has its own strengths and weaknesses. 

> If you are interested in supporting fetching large number of results, may be we can get inspiration from https://www.elastic.co/guide/en/elasticsearch/reference/6.8/search-request-scroll.html, I could research it.

It requires that you store copies of the data in a temporary table. A similar idea was used in Kubernetes: https://github.com/kubernetes/community/pull/896/files
I'm afraid it's too complicated in our case.

**Comment by svetli-n** [comment]
@mik-laj @ephraimbuddy 

setting `maximum: 100` for `PageLimit` results in validation error and `response.status_code == 400` when `/api/v1/connections?limit=150` and same error with e.g. `/api/v1/connections?limit=0` due to `minimum: 1`

Do we want that behaviour (not IMO) or be lenient and set limit=100 (default) for both requests?

**Comment by ephraimbuddy** [comment]
The 400 for the limit=0 LGTM. I will investigate now for the maximum

**Comment by ephraimbuddy** [comment]
I didn't have any issue when I added maximum. You should add it immediately after minimum.
See https://swagger.io/docs/specification/data-models/data-types/

File: airflow/api_connexion/openapi/v1.yaml [hunk 1]
```diff
@@ -2134,6 +2134,7 @@ components:
       schema:
         type: integer
         minimum: 1
+        maximum: 100
```
Comment (mik-laj) [inline airflow/api_connexion/openapi/v1.yaml:2137]: I have introduced a change that will facilitate the introduction of a limit that can be defined by the user. 
https://github.com/apache/airflow/pull/9405
In this way, we will not have a fixed value in API and users will be able to adjust this limit according to their requirements.

Comment (svetli-n) [reply airflow/api_connexion/openapi/v1.yaml:2137]: - isn't this supposed to be done with custom validators? https://connexion.readthedocs.io/en/latest/request.html
https://github.com/zalando/connexion/tree/master/examples/swagger2/enforcedefaults

- isn't the intention to not allow the user to call the API with too high limit?

Comment (mik-laj) [reply airflow/api_connexion/openapi/v1.yaml:2137]: If you want, you can do it using your method, but I think my idea seems simpler to do.

The intention is to limit, but the limit depends on the specific deployment. If your server has 64 GB RAM, then you can set a different limit than if your server has only 1 GB.

**Pull request closed without merging**